src/core/datatypes/phylogenetics/ratematrix/AbstractRateMatrix.cpp

```diff
@@ -582,13 +582,29 @@ void AbstractRateMatrix::exponentiateMatrixByScalingAndSquaring(double t,  Trans
         }
     }
 
+    // sanitization part 1 -- ensure all positive entries
+    for(int i=0;i<num_states;i++)
+        for(int j=0;j<num_states;j++)
+            p[i][j] = std::max(p[i][j], 0.0);
+
+
     // now perform the repeated squaring
     TransitionProbabilityMatrix r(num_states);
     for (size_t i = 0; i < s; i++)
     {
         multiplyMatrices(p, p, r);
         p = r;
+    }
+
+    // sanitization part 2 -- ensure rows sum to 1
```

Comment:
To me this looks dangerous because it doesn't show that there is a problem. So the analysis could be happily continuing while there being serious issue. Shouldn't throwing an error be more appropriate?

Reply:
In summary, throwing an error isn't appropriate because round-off errors are not something we can avoid.  OK?

Reply:
Note that this isn't a new idea, I'm just implementing it in a new place.  This is the solution that we have used previously, it is just that RevBayes has many duplication versions of this code, and they don't all have the sanitization implemented.

Reply:
If we throw an error in the likelihood calculation, would that cancel an MCMC run, or would it in fact just force the proposal to be rejected?  Cancelling the MCMC run wouldn't solve the problem that @jembrown had, but forcing the proposal to be rejected might work.

One possibility would be to renormalize matrix rows if the total sum for each row is in `[0.99, 1.01]`, and force the proposal to be rejected otherwise.  We could do this by 
* setting all transition matrix entries to 0 -- maybe?
* throwing an exception and forcing proposals that throw an exception in the likelihood to be rejected
* other ideas?